Hash api tokens

malakai97 opened the pull request:
Resolves PFDR-106

Relies on the PR for keycard 0.3.0 being merged.

- This does not alter the way that we use keycard's primary authentication mechanisms.
- I'm particularly interested in a way to update the user columns without iterating over all of them individually.

botimer commented on app/controllers/application_controller.rb:
```diff
@@ -50,8 +50,8 @@ def authenticate
-      if (@current_user = User.find_by(api_key: token))
-        @current_user.identity = UserAttributes.new(username: @current_user.username)
+      if (@current_user = User.find_by(api_key_digest: Keycard::ApiKey.new(token).digest))
+        @current_user.identity = UserAttributes.new(username: @current_user.username).all
```
I want to check with @aelkiss on this, but I'm not sure we need the JWT support at all. We could definitely support JWT, but it seems to me that the built-in support is much more in line with browser workflows with third party / social sign-in, neither of which seems relevant here.

aelkiss replied:
No, there's no JWT support required, but I'm not sure exactly what this comment is referring to in the context of the code.

botimer replied:
Per our discussion this morning, merging as-is will cause no trouble, but the Rails-provided authenticate_with_http_token is not really needed. We should trim this down and simplify.

botimer commented on Gemfile:
```diff
@@ -23,8 +23,8 @@ gem "rails", "~> 5.1.0"
-gem "checkpoint", github: "mlibrary/checkpoint"
-gem "keycard", github: "mlibrary/keycard"
+gem "checkpoint", '~> 1.0.3', github: 'mlibrary/checkpoint', tag: 'v1.0.3'
+gem "keycard", "~> 0.3.0", github: 'mlibrary/keycard', branch: 'PFDR-106/api_key'
```
I just merged and released 0.3.0, so this should be changed to the tag.

I guess we can just use the tags if we have to keep these as github sources... I seem to remember being able to use a plain version number _and_ `bundle config local.GEMNAME` somehow, but it eludes me today.

botimer commented on app/controllers/application_controller.rb:
```diff
@@ -60,8 +61,9 @@ def authenticate_token
-    @current_user.identity = Keycard::RequestAttributes.new(request)
-    if @current_user.identity[:username].empty?
+    @current_user.identity = Keycard::Request::AttributesFactory.new.for(request).identity
+    @current_user.identity[:username] = @current_user.identity[:user_pid]
```
This won't do anything for token auth (the user_pid is not set). We should trace what the dependency on `:username` in the identity is -- recent Keycard does not use that key at all, so I think it's vestigial.

botimer commented on app/controllers/application_controller.rb:
```diff
@@ -50,8 +50,9 @@ def authenticate
-        @current_user.identity = UserAttributes.new(username: @current_user.username)
+      digest = Keycard::DigestKey.new(key: token).digest
+      if (@current_user = User.find_by(api_key_digest: digest))
+        @current_user.identity = UserAttributes.new(username: @current_user.username).all
```
Not sure why GitHub won't let me comment on the line above (44) that will need to change. We should just always go to Keycard, since we are planning to use the Authorization header. What I think that really means is that this method goes away, and the Keycard implementation get inlined in the `authenticate` method with this User find for setting the current_user.

botimer commented on app/models/user.rb:
```diff
@@ -6,17 +6,20 @@ class User < ApplicationRecord
+      Keycard::ApiKey.new
+    else
+      Keycard::ApiKey::Hidden.new
+    end
```
@malakai97 tried it; we decided not to do this now.

botimer commented on db/migrate/20190301184824_rename_user_api_key.rb:
```diff
@@ -0,0 +1,19 @@
+    User.all.each do |user|
+      user.update_column(
+        :api_key_digest,
+        Keycard::ApiKey.new(user.read_attribute(:api_key)).digest
```
DigestKey now, double check parameters as our options have recently changed.

botimer commented on lib/user_attributes_2.rb:
```diff
@@ -0,0 +1,19 @@
+# frozen_string_literal: true
```
Extraneous file.

botimer commented on spec/controllers/application_controller_spec.rb:
```diff
@@ -58,9 +58,14 @@ def not_found
+      # TODO: This all goes away once we take the authenticate_with_http_token
+      # built-in out. We have to explicitly access the user, even when defined with let!,
+      # because it is not saved before the controller method executes otherwise... shrug
+      let(:auth_header) { user; { "Authorization" => "Token token=#{key.to_s}" } }
```
Whoops. I left my rough comment in here. It's not really that it "all" goes away since we are planning to use the Authorization header.. What _might_ go away is the immediate access of the user to coerce the timing vs. controller action. I could not figure out why it was necessary here but not in other spots.

malakai97 commented:
- removed todo comment
- removed extraneous file
- fixed and tested migration
- changed gemfile

The application_controller comments seem pending PFDR-107 and like they should be part of PFDR-108.

botimer commented:
Yes, sort of. This code (`authenticate_keycard`) cannot execute if we use the Authorization header (it will always go to `authenticate_token`). I think these changes are intertwined. I would probably address both in the same pull request, but if merging this would help move it along, go for it.
